src/algebra/big_operators/order.lean

```diff
@@ -308,3 +309,62 @@ open opposite
 (@unop_add_hom β _).map_sum _ _
 
 end with_top
+
+namespace finset
+
+/--
+The strong pigeonhole principle for finitely many pigeons and
+pigeonholes: there is a pigeonhole with at least as many pigeons as
+the ceiling of the average number of pigeons across all pigeonholes.
+("The maximum is at least the mean" specialized to integers.)
+
+More formally, given a function between finite sets `s` and `t` and
+setting `n` so that `n + 1` is at most the ceiling of `s.card / t.card`,
+then there is an element of `t` whose preimage contains more
+than `n` elements.  (We formulate the constraint on `n` as
+`t.card * n < s.card`, which incorporates that `s` and `t` must be nonempty.)
+
+See also: `finset.pigeonhole`
+-/
+lemma strong_pigeonhole {s : finset α} {t : finset β} [decidable_eq β]
+  (f : α → β) (hf : ∀ a ∈ s, f a ∈ t)
+  (n : ℕ) (hn : t.card * n < s.card) :
+  ∃ y ∈ t, n < (s.filter (λ x, f x = y)).card :=
+begin
+  classical, by_contra hz, push_neg at hz,
+  suffices h : s.card ≤ t.card * n,
+  have h' := lt_of_le_of_lt h hn,
+  exact nat.lt_asymm h' h',
+
+  have key := λ (y : β) (yel : y ∈ t), (preimage_card_ne_zero_iff_mem_image f hf y).mp,
+
+  calc s.card = ∑ y in s.image f, (s.filter (λ x, f x = y)).card :
+    by apply card_eq_sum_card_image
+          ... = ∑ y in t, (s.filter (λ x, f x = y)).card :
+    by { rw ←sum_filter_of_ne key, congr, convert (filter_mem_image_eq_image f s t hf).symm }
+          ... ≤ ∑ y in t, n :
+    by { convert sum_le_sum hz, simp, }
+          ... = t.card * n :
+    by { simp only [nat.cast_id, nsmul_eq_mul, sum_const] },
+end
+
+end finset
+
+namespace fintype
+open finset
+
+/--
+The strong pigeonhole principle for finitely many pigeons and pigeonholes.
+This is the `fintype` version of `finset.strong_pigeonhole`.
```

Comment:
Can you copy the longer description from `finset.strong_pigeonhole` here, to keep this self-contained?

Reply:
Done